File: src/main/java/hes/sudoku/LogicSolver.java
package hes.sudoku;

import java.util.*;

public class LogicSolver {

    private final Puzzle puzzle;
    private Printer printer;

    public LogicSolver(Puzzle puzzle) {
        this(puzzle, new Printer(puzzle));
    }

    public LogicSolver(Puzzle puzzle, Printer printer) {
        this.puzzle = puzzle;
        this.printer = printer;
    }

    public void solve() {
        while (fillNumbers() || eliminateCandidates());
    }

    public boolean fillNumbers() {
        Set<Move> moves;
        do {
            moves = new LinkedHashSet();

            moves.addAll(findUniqueCandidatesInGroup());
            moves.addAll(findUniqueCandidates());

            applyMoves(moves);

        } while (!moves.isEmpty());

        printer.println("No more moves found.");

        return !moves.isEmpty();
    }

    public boolean eliminateCandidates() {
        Set<Move> eliminations;
        do {
            eliminations = new LinkedHashSet();

            eliminations.addAll(eliminateCrossGroupDependencies());
            eliminations.addAll(eliminateBasedOnUniqueSetsInGroup());

            applyEliminations(eliminations);

        } while (!eliminations.isEmpty());

        printer.println("No more eliminations found.");

        return !eliminations.isEmpty();
    }

    private void applyMoves(Set<Move> moves) {
        for (Move move : moves) {
            printer.println(move);
            puzzle.apply(move);
        }
    }

    private void applyEliminations(Set<Move> eliminations) {
        if (!eliminations.isEmpty()) {
            printer.println("\nEliminating candidates:");
        }

        for (Move elimination : eliminations) {
            printer.println(elimination);
            Cell cell = puzzle.getCell(elimination.cell().getColumn(), elimination.cell().getRow());
            cell.eliminate(elimination.number());
        }
    }


    //
    // Find candidates
    //

    private Set<Move> findUniqueCandidates() {
        Set<Move> moves = new LinkedHashSet();
        for (Cell cell : puzzle.getCells()) {
            if (cell.getCandidates().size() == 1) {
                moves.add(new Move(cell, getCandidate(cell), "No other candidate for this cell"));
            }
        }
        return moves;
    }

    public Integer getCandidate(Cell cell) {
        return cell.getCandidates().iterator().next();
    }


    private Set<Move> findUniqueCandidatesInGroup() {
        Set<Move> candidates = new LinkedHashSet();
        for (Group group : puzzle.getGroups()) {
            candidates.addAll(findGhostCandidates(group));
        }

        return candidates;
    }

    public Set<Move> findGhostCandidates(Group group) {

        Set<Move> candidates = new LinkedHashSet();
        for (int number = 1; number < 10; number++) {
            int nrOfCandidates = 0;
            for (Cell cell : group.getCells()) {
                if (cell.hasCandidate(number)) {
                    nrOfCandidates++;
                }
            }
            if (nrOfCandidates == 1) {
                Cell cell = getFirstGhost(number, group.getCells());
                candidates.add(new Move(cell, number, String.format("Only possibility in %s", group)));
            }
        }

        return candidates;
    }

    private Cell getFirstGhost(Integer number, Collection<Cell> cells) {
        for (Cell cell : cells) {
            if (cell.hasCandidate(number)) {
                return cell;
            }
        }

        throw new IllegalStateException(String.format("No ghost candidate %s in %s", number, cells));
    }


    //
    // Eliminate
    //

    private Set<Move> eliminateCrossGroupDependencies() {
        Set<Move> eliminations = new LinkedHashSet();

        for (Group group : puzzle.getGroups()) {
            for (int number = 1; number <= 9; number++) {
                eliminateCrossGroupDependencies(group, number, puzzle.getBoxes(), eliminations);
                eliminateCrossGroupDependencies(group, number, puzzle.getRows(), eliminations);
                eliminateCrossGroupDependencies(group, number, puzzle.getColumns(), eliminations);
            }
        }

        return eliminations;
    }

    private void eliminateCrossGroupDependencies(Group group, Integer number, List<Group> ignore, Set<Move> eliminations) {

        Set<Group> overlapping = getOverlappingGroupsWithSameNumber(group, number);
        overlapping.removeAll(ignore);

        if (overlapping.size() == 1) {
            Group other = overlapping.iterator().next();
            Set<Cell> cells = new LinkedHashSet<>(other.getCells());
            cells.removeAll(group.getCells());

            for (Cell cell: cells) {
                if (!cell.hasCandidate(number)) {
                    continue;
                }
                eliminations.add(new Move(cell, number, String.format("Eliminating %s because it needs to be in %s, %s", number, group, other)));
            }
        }
    }

    private Set<Group> getOverlappingGroupsWithSameNumber(Group group, Integer number) {
        Set<Group> overlapping = new HashSet<>();
        for (Cell cell : group.getCells()) {
            if (cell.hasCandidate(number)) {
                overlapping.addAll(cell.getGroups());
            }
        }
        overlapping.remove(group);

        return overlapping;
    }

    //
    //
    //

    public Set<Move> eliminateBasedOnUniqueSetsInGroup() {
        Set<Move> eliminations = new LinkedHashSet();

        for (Group group : puzzle.getGroups()) {
            eliminateBasedOnUniqueSets(group, eliminations);
        }

        return eliminations;
    }

    private static void eliminateBasedOnUniqueSets(Group group, Set<Move> eliminations) {
        Map<Set<Integer>, Integer> candidateSetCount = new LinkedHashMap<>();
        for (Cell cell : group.getCells()) {
            add(candidateSetCount, cell.getCandidates());
        }

        for (Map.Entry<Set<Integer>, Integer> entry : candidateSetCount.entrySet()) {
            if (entry.getKey().size() == entry.getValue()) {
                removeOtherDependencies(group, entry.getKey(), eliminations);
            }
        }
    }

    private static void removeOtherDependencies(Group group, Set<Integer> set, Set<Move> eliminations) {
        for (Cell cell : group.getCells()) {
            if (!cell.getCandidates().equals(set)) {
                for (Integer number : set) {
                    if (cell.hasCandidate(number)) {
                        eliminations.add(new Move(cell, number, String.format("Eliminate based on unique set in %s", group)));
                    }
                }
            }
        }
    }

    private static <T> void add(Map<T, Integer> count, T something) {
        if (!count.containsKey(something)) {
            count.put(something, 1);
        } else {
            count.put(something, count.get(something) + 1);
        }
    }
}

File: src/main/java/hes/sudoku/kt/LogicSolver.kt
package hes.sudoku.kt

class LogicSolver(
        val puzzle: Puzzle,
        private val printer: Printer = DefaultPrinter(puzzle)) {

    fun solve() {
        while (fillNumbers() || eliminateCandidates());
    }

    fun fillNumbers(): Boolean {
        var moves: MutableSet<Move>
        do {
            moves = mutableSetOf()
            moves.addAll(findUniqueCandidatesInGroup())
            moves.addAll(findUniqueCandidates())
            applyMoves(moves)
        } while (!moves.isEmpty())

        printer.println("No more moves found.")

        return !moves.isEmpty()
    }

    private fun eliminateCandidates(): Boolean {
        var eliminations: MutableSet<Move>
        do {
            eliminations = mutableSetOf()
            eliminations.addAll(eliminateCrossGroupDependencies())
            eliminations.addAll(eliminateBasedOnUniqueSetsInGroup())
            applyEliminations(eliminations)
        } while (!eliminations.isEmpty())

        printer.println("No more eliminations found.")

        return !eliminations.isEmpty()
    }

    private fun applyMoves(moves: Set<Move>) {
        for (move in moves) {
            printer.println(move)
            puzzle.apply(move)
        }
    }

    private fun applyEliminations(eliminations: Set<Move>) {
        if (!eliminations.isEmpty()) {
            printer.println("\nEliminating candidates:")
        }

        for (removal in eliminations) {
            printer.println(removal)
            puzzle.getCell(removal.cell.column, removal.cell.row)
                    .eliminate(removal.number)
        }
    }

    //
    // Find candidates
    //

    private fun findUniqueCandidates(): Set<Move> {
        return puzzle.cells.filter { it.candidates.size == 1 }
                .mapTo(mutableSetOf()) { cell ->
                    Move(cell, getCandidate(cell)).withDescription("No other candidate for this cell")
        }
    }

    fun getCandidate(cell: Cell): Int {
        return cell.candidates.first()
    }

    private fun findUniqueCandidatesInGroup(): Set<Move> {
        val candidates = mutableSetOf<Move>()
        for (group in puzzle.groups) {
            candidates.addAll(findGhostCandidates(group))
        }
        return candidates
    }

    fun findGhostCandidates(group: Group): Set<Move> {
        val candidates = mutableSetOf<Move>()
        for (number in 1..9) {
            var nrOfCandidates = 0
            for (cell in group.cells) {
                if (number in cell.candidates) {
                    nrOfCandidates++
                }
            }
            if (nrOfCandidates == 1) {
                val cell = getFirstGhost(number, group.cells)
                candidates.add(Move(cell, number).withDescription("Only possibility in $group"))
            }
        }
        return candidates
    }

    private fun getFirstGhost(number: Int, cells: Collection<Cell>): Cell {
        return cells.filter { number in it.candidates }.first()
    }

    //
    // Eliminate
    //

    private fun eliminateCrossGroupDependencies(): Set<Move> {
        val eliminations = mutableSetOf<Move>()
        for (group in puzzle.groups) {
            for (number in 1..9) {
                eliminateCrossGroupDependencies(group, number, puzzle.boxes, eliminations)
                eliminateCrossGroupDependencies(group, number, puzzle.rows, eliminations)
                eliminateCrossGroupDependencies(group, number, puzzle.columns, eliminations)
            }
        }
        return eliminations
    }

    private fun eliminateCrossGroupDependencies(group: Group, number: Int, ignore: List<Group>, eliminations: MutableSet<Move>) {
        val overlapping = getOverlappingGroupsWithSameNumber(group, number)
        overlapping.removeAll(ignore)
        if (overlapping.size == 1) {
            val other = overlapping.iterator().next()
            val cells = other.cells.toMutableSet()
            cells.removeAll(group.cells)
            for (cell in cells.filter { number in it.candidates }) {
                eliminations.add(Move(cell, number).withDescription("Eliminating $number because it needs to be in $group, $other"))
            }
        }
    }

    private fun getOverlappingGroupsWithSameNumber(group: Group, number: Int): MutableSet<Group> {
        val overlapping = mutableSetOf<Group>()
        for (cell in group.cells.filter { number in it.candidates }) {
            overlapping.addAll(cell.groups)
        }
        overlapping.remove(group)

        return overlapping
    }

    //
    // Eliminate candidates in a group when there is a subset of n candidates occurring n times.
    // The elements of that particular set can't be candidates anywhere else in that group then.
    // For example, you have a row that have two places with {2,4} as candidates. Then another cell
    // with {2, 4, 6} should not have {2, 4}.
    //

    fun eliminateBasedOnUniqueSetsInGroup(): Set<Move> {
        val eliminations: MutableSet<Move> = LinkedHashSet()
        for (group in puzzle.groups) {
            eliminateBasedOnUniqueSets(group, eliminations)
        }

        return eliminations
    }
}

fun eliminateBasedOnUniqueSets(group: Group, eliminations: MutableSet<Move>) {
    val candidateSetCount: MutableMap<Set<Int>, Int> = LinkedHashMap()
    for (cell in group.cells) {
        add(candidateSetCount, cell.candidates)
    }

    for ((key, value) in candidateSetCount) {
        if (key.size == value) {
            removeOtherDependencies(group, key, eliminations)
        }
    }
}

private fun removeOtherDependencies(group: Group, set: Set<Int>, eliminations: MutableSet<Move>) {
    for (cell in group.cells.filter { cell -> cell.candidates != set }) {
        for (number in set.filter { number -> number in cell.candidates }) {
            eliminations.add(Move(cell, number).withDescription("Eliminate $number from $cell, because it is in a unique set $set elsewhere in $group"))
        }
    }
}

private fun <T> add(count: MutableMap<T, Int>, something: T) {
    if (!count.containsKey(something)) {
        count[something] = 1
    } else {
        count[something] = count[something]!! + 1
    }
}
